### ISSUE TITLE
[HOLD 25935][$1000] Chat - 'New message' pill appears after deleting unread message

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#22174]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___
**Issue found when executing PR** https://github.com/Expensify/App/pull/21557
## Action Performed:
1. Go to staging.new.expensify.com
2. Go to any chat
3. Send a message
4. Mark the message unread
5. Scroll up and click on the 'new message' pill
6. Delete the unread message
7. Scroll up

## Expected Result:
'New message' pill will not appear as the unread message is deleted

## Actual Result:
'New message' pill still appears although the unread message is deleted
If the 'new message' pill is not clicked, this issue will not occur.

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:**  1.3.36.2

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

![Bug6115722_bandicam_2023-07-04_17-01-06-305](https://github.com/Expensify/App/assets/78819774/d6cd4b30-4b37-4fa3-831b-cd9ca7ae483b)

https://github.com/Expensify/App/assets/78819774/07f29a78-5edf-4d0d-aa42-c9cf1bdfe13c

**Expensify/Expensify Issue URL:**

**Issue reported by:** Applause - Internal Team

**Slack conversation:**

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01c35abc455d11baf9</li>
        <li>Upwork Job ID: 1678381225562419200</li>
        <li>Last Price Increase: 2023-07-31</li>
    </ul>
</details>

**Comment by DanutGavrus** [comment]
This reported issue has the same root cause as [this](https://github.com/Expensify/App/pull/21557#issuecomment-1612865965) which was found while testing [this PR](https://github.com/Expensify/App/pull/21557). We were [thinking](https://github.com/Expensify/App/pull/21557#issuecomment-1614981455) and [decided](https://github.com/Expensify/App/issues/21875#issuecomment-1614996939) that the issue is out of scope of that PR.

Here are my findings:
1. This reported bug happens if we remove all the changes introduced by the PR;
2. This reported bug does **not** happen if we introduce the changes proposed [here](https://github.com/Expensify/App/issues/21875#issuecomment-1613200013).




Right now, *New Markers Issues* are on hold as I've found out from [this comment](https://github.com/Expensify/App/issues/21875#issuecomment-1618774447). Lately I've seen quite some issues arising from the **same root cause**, which is that the *New Markers do not get removed when report becomes read*. Maybe we want to make an exception to fix them all with [1 Proposal](https://github.com/Expensify/App/issues/21875#issuecomment-1613200013), or we'll keep them all on hold until the refactor is done.

**Comment by anmurali** [comment]
<img width="1278" alt="image" src="https://github.com/Expensify/App/assets/14225673/08026e47-80e5-414a-97b6-44044dcd48af">

Marking as unread in staging does not show me a `New message` pill. Not sure how to reproduce this. I am on Chrome/ Mac

**Comment by DanutGavrus** [comment]
Based on [this](https://github.com/Expensify/App/issues/22174#issuecomment-1621345697) I think we'll have to put this on hold for [#15212](https://github.com/Expensify/App/issues/15212)

**Comment by anmurali** [comment]
@0xmiroslav - do you agree with @DanutGavrus? If yes, I will add the held to the title and degrade this to a weekly while we wait.

**Comment by anmurali** [comment]
@0xmiroslav bump

**Comment by DanutGavrus** [comment]
@0xmiroslav 
> ReportActionsView is now refactored to function component. Please update your proposals accordingly.

## Proposal
### Please re-state the problem that we are trying to solve in this issue.
Chat - 'New message' pill appears after deleting unread message.

### What is the root cause of that problem?
1. Clicking on the *New Messages* pill automatically sets the report as *read*;
2. But, in `ReportActionsView.js` we do not check if the report automatically became `read` in order to clear `newMarkerReportActionID`'s value.

### What changes do you think we should make in order to solve the problem?
We should add this logic:
```js
useEffect(() => {
    // If the report automatically became read, remove the *New* green marker and the *New Messages* notification at scroll.
    if (!ReportUtils.isUnread(props.report) && newMarkerReportActionID != '') {
        setNewMarkerReportActionID('');
    }
}, [props.report, newMarkerReportActionID]);
```

**Comment by DanutGavrus** [comment]
@0xmiroslav 
> Please update your proposals accordingly.  

[Proposal](https://github.com/Expensify/App/issues/22174#issuecomment-1640592254) updated. What are your thoughts about it? Thanks!

**Comment by anmurali** [comment]
@0xmiroslav I believe we're waiting on you to review the updated proposal(s)?

**Comment by DanutGavrus** [comment]
@0xmiroslav bump

**Comment by anmurali** [comment]
@0xmiroslav anything?

**Comment by anmurali** [comment]
Bumped [here](https://expensify.slack.com/archives/C049HHMV9SM/p1692110506248709)

**Comment by situchan** [comment]
I will take this as C+

**Comment by DanutGavrus** [comment]
@situchan I'm going to leave a new Proposal as https://github.com/Expensify/App/issues/22174#issuecomment-1640592254 got outdated.

## Proposal

### Please re-state the problem that we are trying to solve in this issue.
New message pill appears after deleting unread message.

### What is the root cause of that problem?
In `ReportActionsList` component, we have this `useEffect` which **clears** the *current unread marker* so that it can be **recalculated**:
https://github.com/Expensify/App/blob/c20c3870ee46d6788de554be27706e9898638904/src/pages/home/report/ReportActionsList.js#L164-L177
1. When deleting the last message, we set `lastVisibleActionCreated: ''` and we leave `lastReadTime` untouched, so the `useEffect` does not trigger;
2. Currently, after deleting the last message, `ReportUtils.isUnread(report)` is `false`, thus we'll enter the **if** which **returns** before *clearing the current unread marker*.

### What changes do you think we should make in order to solve the problem?
1. We should add `report.lastVisibleActionCreated` to the `useEffect` in order to trigger it;
2. We should add `currentUnreadMarker.current = null` when entering the **if** too in order to trigger a **recalculation**.  

After a small refactoring, the updated `useEffect` would look like this:
```js
useEffect(() => {
    // Clearing the current unread marker so that it can be recalculated
    currentUnreadMarker.current = null;

    const didManuallyMarkReportAsUnread = report.lastReadTime < DateUtils.getDBTime() && ReportUtils.isUnread(report);
    if (didManuallyMarkReportAsUnread) {
        setMessageManuallyMarked({read: true});
    } else {
        setMessageManuallyMarked({read: false});
    }

}, [report.lastReadTime, report.lastVisibleActionCreated]);
```

@situchan This would also fix this issue:
1. Mark a message as unread;
2. Scroll up and click on the *new messages* pill;
3. Scroll up again and see that the *new messages* pill is still present no matter how many times you click on it.

**Comment by DanutGavrus** [comment]
Hi, @situchan! Would you have the time to review my above Proposal? Thanks!

**Comment by anmurali** [comment]
Bumped @situchan [here](https://expensify.slack.com/archives/C049HHMV9SM/p1693964557405329)

**Comment by napster125** [comment]
## Proposal
### Please re-state the problem that we are trying to solve in this issue.
New message pill appears after deleting unread message.
### What is the root cause of that problem?
The problem lies within the `useEffect` hook, specifically in this condition block:
https://github.com/Expensify/App/blob/1bdc5acc12f982b4f6e37bfadf9060ac3d2d3b3c/src/pages/home/report/ReportActionsList.js#L142-L162
```
if (currentUnreadMarker.current || reportActionSize.current === sortedReportActions.length)
```
This hook is used to manage the addition and removal of report items. However, during this process, `currentUnreadMarker.current` is not `null`, which results in undesired behavior, preventing the cleaning of `currentUnreadMarker.current`. The condition statement is responsible for managing the display of floating messages for new messages, hence the issue.

To simplify, we use this condition to clear the floating message counter component for new messages.
https://github.com/Expensify/App/blob/1bdc5acc12f982b4f6e37bfadf9060ac3d2d3b3c/src/pages/home/report/ReportActionsList.js#L155-L160
So, in simpler terms, this condition is incorrect.
### What changes do you think we should make in order to solve the problem?
We can easily modify the condition as follows.
```
if (!currentUnreadMarker.current || reportActionSize.current === sortedReportActions.length)
```
This modification ensures that the new message pill no longer appears after deleting an unread message.
### What alternative solutions did you explore? (Optional)
N/A
### Result

https://github.com/Expensify/App/assets/136437976/9858910b-bbb0-4d5b-8dfc-771d8f7a1df0

**Comment by Pluto0104** [comment]
## Proposal
### Please re-state the problem that we are trying to solve in this issue.
New message pill appears after deleting unread message.
### What is the root cause of that problem?
The problem lies within the `useEffect` hook, specifically in this condition block:
https://github.com/Expensify/App/blob/1bdc5acc12f982b4f6e37bfadf9060ac3d2d3b3c/src/pages/home/report/ReportActionsList.js#L142-L162
```
if (currentUnreadMarker.current || reportActionSize.current === sortedReportActions.length)
```
This hook is used to manage the addition and removal of report items. However, during this process, `currentUnreadMarker.current` is not `null`, which results in undesired behavior, preventing the cleaning of `currentUnreadMarker.current`. The condition statement is responsible for managing the display of floating messages for new messages, hence the issue.

To simplify, we use this condition to clear the floating message counter component for new messages.
https://github.com/Expensify/App/blob/1bdc5acc12f982b4f6e37bfadf9060ac3d2d3b3c/src/pages/home/report/ReportActionsList.js#L155-L160
So, in simpler terms, this condition is incorrect.
### What changes do you think we should make in order to solve the problem?
We can easily modify the condition as follows.
```
if (!currentUnreadMarker.current || reportActionSize.current === sortedReportActions.length)
```
This modification ensures that the new message pill no longer appears after deleting an unread message.
### What alternative solutions did you explore? (Optional)
N/A
### Result

https://github.com/Expensify/App/assets/136437976/9858910b-bbb0-4d5b-8dfc-771d8f7a1df0

**Comment by situchan** [comment]
updating today as merge freeze is over

**Comment by Pluto0104** [comment]
Hi @situchan, When you have a moment, could you please take a look at [my proposal](https://github.com/Expensify/App/issues/22174#issuecomment-1709186899)? I'd appreciate your feedback. Thank you!

**Comment by situchan** [comment]
> Hi @situchan, When you have a moment, could you please take a look at [my proposal](https://github.com/Expensify/App/issues/22174#issuecomment-1709186899)? I'd appreciate your feedback. Thank you!

Sure!

**Comment by situchan** [comment]
sorry missed this. will provide update asap

**Comment by anmurali** [comment]
Bumped again in Slack - https://expensify.slack.com/archives/C049HHMV9SM/p1695657360236889?thread_ts=1693964557.405329&cid=C049HHMV9SM

**Comment by situchan** [comment]
sorry was focusing on priority issues. will provide update by this week

**Comment by situchan** [comment]
Sever is very slow at the moment and cannot test New message pill properly.
I will come back to this when it's stable

**Comment by it-education-md** [comment]
## Proposal
### Please re-state the problem that we are trying to solve in this issue.
New message pill appears after deleting unread message.
### What is the root cause of that problem?
The problem lies within the `useEffect` hook, specifically in this condition block:
https://github.com/Expensify/App/blob/1bdc5acc12f982b4f6e37bfadf9060ac3d2d3b3c/src/pages/home/report/ReportActionsList.js#L142-L162
```
if (currentUnreadMarker.current || reportActionSize.current === sortedReportActions.length)
```
This hook is used to manage the addition and removal of report items. However, during this process, `currentUnreadMarker.current` is not `null`, which results in undesired behavior, preventing the cleaning of `currentUnreadMarker.current`. The condition statement is responsible for managing the display of floating messages for new messages, hence the issue.

To simplify, we use this condition to clear the floating message counter component for new messages.
https://github.com/Expensify/App/blob/1bdc5acc12f982b4f6e37bfadf9060ac3d2d3b3c/src/pages/home/report/ReportActionsList.js#L155-L160
So, in simpler terms, this condition is incorrect.
### What changes do you think we should make in order to solve the problem?
We can easily modify the condition as follows.
```
if (!currentUnreadMarker.current || reportActionSize.current === sortedReportActions.length)
```
This modification ensures that the new message pill no longer appears after deleting an unread message.
### What alternative solutions did you explore? (Optional)
N/A
### Result

https://github.com/Expensify/App/assets/136437976/9858910b-bbb0-4d5b-8dfc-771d8f7a1df0

**Comment by it-education-md** [comment]
Contributor details
Your Expensify account email: iteducation.md@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01564d5fb5dd2f7ad1

**Comment by DanutGavrus** [comment]
Hi, @situchan!

I've retested this and it still happens. I've slightly updated [my Proposal](https://github.com/Expensify/App/issues/22174#issuecomment-1700668447) from **31 Aug** again according to the latest changes to the file. Hopefully you'll get the chance to review this soon! Thanks!

**Comment by situchan** [comment]
@DanutGavrus thanks. I will provide feedback today

**Comment by situchan** [comment]
I have a question about the expected behavior after click "New messages" button:

Option 1. scroll down to newest message, keep showing unread indicator and when scroll up again, "New messages" button still shows
Option 2. scroll down to newest message, remove unread indicator and when scroll up again, "New messages" button no longer shows

NOTE: unread message is not deleted

In another word, "New messages" button just updates scroll position or hide unread badges as well?

@Expensify/design which one is correct?

I think Option 1 is correct because it's consistent with "Mark as read" behavior from LHN


https://github.com/Expensify/App/assets/108292595/4f2045b1-aa2f-4572-ae95-f8ccfaffc8d3

**Comment by DanutGavrus** [comment]
@situchan 
I think that **Option 1** looks like a bug for the user. **Option 2** will keep consistency between pressing on **New Messages** and marking the report as read from the LHN by implementing [this Proposal](https://github.com/Expensify/App/issues/22174#issuecomment-1700668447)(Marking as read from LHN will also remove the new marker indicators inside the chat).

From what I see, we'll also want to remove the **New Messages** floating indicator after pressing on it: https://github.com/Expensify/App/issues/21288.

And we'll also want to remove the **New Messages** floating indicator after marking reports as read from the LHN: https://github.com/Expensify/App/issues/21453.

That Proposal would fix all 3 Issues.

**Comment by situchan** [comment]
@DanutGavrus no, it's intentional to still show the unread indicator when read report (click from LHN) so user knows what messages they unread yet.

**Comment by DanutGavrus** [comment]
@situchan So if you have unread messages(or mark any as unread) in a report and then mark the report as read, you'll still have the same unread messages in the report? That's a strange behaviour for a user IMO, but if it's desired, we'll surely need another Proposal here.

**Comment by it-education-md** [comment]
Hi, @situchan My solution would fix this issue effectively without changing other logics.

**Comment by situchan** [comment]
I'd like to get confirmation from engineer on https://github.com/Expensify/App/issues/22174#issuecomment-1752227387 as it's more like logic discussion than design discussion.
🎀 👀 🎀

**Comment by MonilBhavsar** [comment]
Option 1 is correct based on the design doc 👍
Just to note, there's an ongoing PR to update the unread functionality. May be if we want to put this issue on HOLD for that, so we don't update code in parallel https://github.com/Expensify/App/pull/25935

**Comment by amyevans** [comment]
Okay thanks so much @MonilBhavsar!

**Comment by amyevans** [comment]
Still on hold

**Comment by situchan** [comment]
Same

**Comment by amyevans** [comment]
Still holding

**Comment by MonilBhavsar** [comment]
This issue should be fixed! We can retest and most probably close this

**Comment by amyevans** [comment]
Thanks Monil! Retested and confirmed it's fixed:

https://github.com/Expensify/App/assets/7277067/129e1f5a-a165-4e17-93fe-717a9d503180
